opengovsg/FormSG
# feat: enable global beta features through DB

## Problem


Manually adding beta flags for individual users is unnecessarily tedious in non-production environments.

## Solution


Create a new DB collection to store global beta flag statuses.

**Breaking Changes** 


**Features**:

- Create new `featureflags` DB collection
- Show beta features if either user or global beta flag is set to enabled; do not show beta features if both flags are not enabled

**Bug Fixes**:

- Rename `add-payment-flag.js` and `multi-language-stats` scripts' directory 

## Tests


Currently, the only beta feature is payments, so we will be basing the tests on the payment feature.

1. Ensure that users with no beta flag + no global beta flag enabled will not see beta features:
- [ ] On the DB, check that:
   - [ ] In the `user` collection, your user document, `betaFlags: { payment: false }`.
   - [ ] In the `globalBeta` collection, a document has the following values `{ name: payment, enabled: false }`.
- [ ] Create a new storage mode form. Go to the settings tab of a storage form, there should not be a "Payments" drawer.
- [ ] Go to the create tab of a storage form, there should not be a "Payments" tab in the builder drawer.
- [ ] Payment preview should not be shown in:
   - [ ] Builder view
   - [ ] Admin preview
   - [ ] Public form

2. Ensure that all users will see beta feature if global beta flag is enabled:
- [ ] On the DB, create the `globalBeta` collection if it doesn't exist. Add a document `{ name: payment, enabled: true }`.
- [ ] On the DB, go to the `user` collection, find your user document, and set `payment` beta flag to `false`.
- [ ] Go to the settings tab of a storage form, there should be a "Payments" drawer.
   - [ ] Connect to a Stripe account.
- [ ] Go to the create tab of a storage form, there should be a "Payments" tab in the builder drawer.
   - [ ] Enable payments, edit name and payment amount. Save.
   - [ ] Payment preview should reflect saved changes accordingly.
- [ ] Form preview should show payment preview accordingly.
- [ ] Public form should:
   - [ ] Show payment preview accordingly.
   - [ ] Allow payment successfully.

3. Ensure that users with user beta flag will still see beta feature if global beta flag is enabled:
- [ ] On the DB, go to the `globalBeta` collection, find the document with name `payment` and make sure the `enabled` value is `true`.
- [ ] On the DB, go to the `user` collection, find your user document, and set `payment` beta flag to `true`.
- [ ] Go to the settings tab of a storage form, there should be a "Payments" drawer.
   - [ ] Connect to a Stripe account.
- [ ] Go to the create tab of a storage form, there should be a "Payments" tab in the builder drawer.
   - [ ] Enable payments, edit name and payment amount. Save.
   - [ ] Payment preview should reflect saved changes accordingly.
- [ ] Form preview should show payment preview accordingly.
- [ ] Public form should:
   - [ ] Show payment preview accordingly.
   - [ ] Allow payment successfully.

4. Ensure that if global beta flag is not enabled, users with user beta flag will still see beta feature:
- [ ] On the DB, go to the `globalBeta` collection, find the document with name `payment` and set the `enabled` value to `false`.
- [ ] On the DB, go to the `user` collection, find your user document, and set `payment` beta flag to `true`.
- [ ] Go to the settings tab of a storage form, there should be a "Payments" drawer.
   - [ ] Connect to a Stripe account.
- [ ] Go to the create tab of a storage form, there should be a "Payments" tab in the builder drawer.
   - [ ] Enable payments, edit name and payment amount. Save.
   - [ ] Payment preview should reflect saved changes accordingly.
- [ ] Form preview should show payment preview accordingly.
- [ ] Public form should:
   - [ ] Show payment preview accordingly.
   - [ ] Allow payment successfully.

## Deploy Notes



**New scripts**:

- `scripts/20230508_add-feature-flag/add-feature-flag-collection.js`: run in Studio3T for new `featureflags` collection to be added
